File: python/examples/basic_usage.py
#!/usr/bin/env python3
"""
Basic usage examples for NodeMaven proxy API.
Demonstrates core functionality including user info, countries, and proxy usage.
"""

import os
import sys
from typing import Optional

# Add parent directory to path for imports
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

try:
    from nodemaven.client import NodeMavenClient
    from nodemaven.utils import get_proxy_config, format_bytes, get_current_ip
    from nodemaven.exceptions import NodeMavenAPIError
except ImportError as e:
    print(f"❌ Import Error: {e}")
    print("💡 Make sure you're in the python/ directory and have the nodemaven package")
    sys.exit(1)


def test_proxy_connection(description: str, proxy_config: dict) -> Optional[str]:
    """Test proxy connection and return IP address."""
    try:
        # Use our IP checking function
        ip = get_current_ip(proxies=proxy_config, timeout=15)
        if ip:
            print(f"   ✅ {description} working! Your IP: {ip}")
            return ip
        else:
            print(f"   ❌ {description} failed to get IP")
            return None
    except Exception as e:
        print(f"   ❌ {description} error: {e}")
        return None


def main():
    """Demonstrate basic NodeMaven usage."""
    print("🚀 NodeMaven Basic Usage Example")
    print("=" * 50)
    
    try:
        # Initialize client
        client = NodeMavenClient()
        print("✅ Client initialized successfully")
        
        # Get user information
        print("\n📊 Getting user information...")
        user_info = client.get_user_info()
        
        print(f"   📧 Email: {user_info.get('email', 'Unknown')}")
        print(f"   👤 Proxy Username: {user_info.get('proxy_username', 'Unknown')}")
        print(f"   🔑 Proxy Password: {user_info.get('proxy_password', 'Unknown')[:8]}...")
        
        # Show data usage
        traffic_used = user_info.get('traffic_used', 0)
        traffic_limit = user_info.get('traffic_limit', 0)
        if traffic_limit > 0:
            remaining = max(0, traffic_limit - traffic_used)
            print(f"   📈 Data Remaining: {format_bytes(remaining)}")
        
        subscription = user_info.get('subscription_type') or user_info.get('subscription')
        print(f"   🎯 Subscription: {subscription}")
        
        # Get available countries
        print("\n🌍 Getting available countries...")
        countries = client.get_countries(limit=10)
        
        if countries.get('results'):
            print(f"   Found {len(countries['results'])} countries (showing first 10):")
            for country in countries['results'][:10]:
                print(f"   - {country.get('name', 'Unknown')} ({country.get('code', 'Unknown')})")
        
        # Get regions for a specific country (US)
        print("\n🏞️  Getting regions for United States...")
        try:
            regions = client.get_regions(country_code="US", limit=5)
            if regions.get('results'):
                print(f"   Found {len(regions['results'])} regions")
        except Exception as e:
            print(f"   ⚠️  Could not get regions: {e}")
        
        # Test basic proxy usage
        print("\n🔌 Testing basic proxy usage...")
        basic_proxies = get_proxy_config()
        test_proxy_connection("Basic proxy", basic_proxies)
        
        # Test geo-targeted proxy
        print("\n🎯 Testing geo-targeted proxy (US)...")
        us_proxies = get_proxy_config(country="US")
        test_proxy_connection("US geo-targeted proxy", us_proxies)
        
        # Test sticky session
        print("\n📌 Testing sticky session...")
        session_id = "example_session_123"
        session_proxies = get_proxy_config(country="US", session=session_id)
        
        # Make multiple requests with same session
        ips = []
        for i in range(3):
            ip = get_current_ip(proxies=session_proxies, timeout=10)
            if ip:
                print(f"   ✅ Sticky session request {i+1} working! Your IP: {ip}")
                ips.append(ip)
            else:
                print(f"   ❌ Sticky session request {i+1} failed")
        
        # Check if all IPs are the same (sticky session working)
        if len(set(ips)) == 1 and ips:
            print(f"   ✅ Sticky session working! All requests used same IP: {ips[0]}")
        elif ips:
            print(f"   ⚠️  Sticky session may not be working - got {len(set(ips))} different IPs")
        
        print("\n" + "=" * 50)
        print("🎉 Example completed!")
        
    except NodeMavenAPIError as e:
        print(f"❌ NodeMaven API Error: {e}")
    except Exception as e:
        print(f"❌ Unexpected Error: {e}")


if __name__ == "__main__":
    main() 

File: php/examples/basic_usage.php
<?php

/**
 * NodeMaven PHP SDK - Basic Usage Example
 * 
 * This example demonstrates the core functionality of the NodeMaven PHP SDK
 */

declare(strict_types=1);

require_once __DIR__ . '/../vendor/autoload.php';

use NodeMaven\Client;
use NodeMaven\Utils;
use NodeMaven\Exceptions\NodeMavenException;

// Load environment variables
Utils::loadEnv(__DIR__ . '/../.env');

echo "🚀 NodeMaven PHP SDK - Basic Usage Example\n";
echo str_repeat("=", 60) . "\n\n";

try {
    // Initialize the client
    echo "1️⃣  Initializing NodeMaven client...\n";
    $client = new Client();
    echo "✅ Client initialized successfully!\n\n";

    // Get user information
    echo "2️⃣  Getting user information...\n";
    $userInfo = $client->getUserInfo();
    
    echo "✅ User Information:\n";
    echo "   Username: " . ($userInfo['username'] ?? $userInfo['email'] ?? 'N/A') . "\n";
    echo "   Proxy Username: " . ($userInfo['proxy_username'] ?? 'N/A') . "\n";
    
    if (isset($userInfo['traffic_used'], $userInfo['traffic_limit'])) {
        echo "   Traffic Used: " . Utils::formatBytes($userInfo['traffic_used']) . "\n";
        echo "   Traffic Limit: " . Utils::formatBytes($userInfo['traffic_limit']) . "\n";
        $remaining = max(0, $userInfo['traffic_limit'] - $userInfo['traffic_used']);
        echo "   Remaining: " . Utils::formatBytes($remaining) . "\n";
    }
    echo "\n";

    // Get available countries
    echo "3️⃣  Getting available countries...\n";
    $countries = $client->getCountries(['limit' => 10]);
    
    echo "✅ Available Countries (first 10):\n";
    foreach ($countries['results'] ?? [] as $index => $country) {
        echo "   " . ($index + 1) . ". {$country['name']} ({$country['code']})\n";
    }
    echo "\n";

    // Basic proxy configuration
    echo "4️⃣  Creating basic proxy configuration...\n";
    $proxyConfig = $client->getProxyConfig();
    
    echo "✅ Basic Proxy Configuration:\n";
    echo "   Host: " . $proxyConfig->getHost() . "\n";
    echo "   HTTP Port: " . $proxyConfig->getHttpPort() . "\n";
    echo "   SOCKS5 Port: " . $proxyConfig->getSocks5Port() . "\n";
    echo "   Username: " . $proxyConfig->getUsername() . "\n";
    echo "   HTTP URL: " . $proxyConfig->getHttpUrl() . "\n";
    echo "\n";

    // Geo-targeted proxy configuration
    echo "5️⃣  Creating geo-targeted proxy (US)...\n";
    $usProxyConfig = $client->getProxyConfig(['country' => 'US']);
    
    echo "✅ US Proxy Configuration:\n";
    echo "   Username: " . $usProxyConfig->getUsername() . "\n";
    echo "   SOCKS5 URL: " . $usProxyConfig->getSocks5Url() . "\n";
    echo "\n";

    // Sticky session example
    echo "6️⃣  Creating sticky session proxy...\n";
    $sessionId = 'my_session_' . Utils::generateSessionId();
    $stickyProxyConfig = $client->getProxyConfig([
        'country' => 'US',
        'session' => $sessionId
    ]);
    
    echo "✅ Sticky Session Proxy:\n";
    echo "   Session ID: {$sessionId}\n";
    echo "   Username: " . $stickyProxyConfig->getUsername() . "\n";
    echo "\n";

    // Test proxy connection
    echo "7️⃣  Testing proxy connection...\n";
    $testResult = $usProxyConfig->testConnection('https://httpbin.org/ip');
    
    if ($testResult['success']) {
        echo "✅ Proxy Connection Test:\n";
        echo "   Status: Success\n";
        echo "   IP Address: " . $testResult['ip'] . "\n";
        echo "   Status Code: " . $testResult['status_code'] . "\n";
    } else {
        echo "❌ Proxy Connection Test Failed:\n";
        echo "   Error: " . $testResult['error'] . "\n";
    }
    echo "\n";

    // Using proxy with Guzzle
    echo "8️⃣  Making HTTP request through proxy...\n";
    $httpClient = $usProxyConfig->createHttpClient();
    
    try {
        $response = $httpClient->get('https://httpbin.org/json');
        $data = json_decode($response->getBody()->getContents(), true);
        
        echo "✅ HTTP Request through Proxy:\n";
        echo "   Status Code: " . $response->getStatusCode() . "\n";
        echo "   Response contains slideshow: " . (isset($data['slideshow']) ? 'Yes' : 'No') . "\n";
    } catch (Exception $e) {
        echo "❌ HTTP Request Failed: " . $e->getMessage() . "\n";
    }
    echo "\n";

    // Multiple proxy configurations
    echo "9️⃣  Creating multiple proxy configurations...\n";
    $countries = ['US', 'UK', 'CA', 'AU', 'DE'];
    
    foreach ($countries as $country) {
        $proxy = $client->getProxyConfig(['country' => $country]);
        echo "   {$country}: " . $proxy->getHost() . ":" . $proxy->getHttpPort() . "\n";
    }
    echo "\n";

    echo str_repeat("=", 60) . "\n";
    echo "🎉 Basic usage example completed successfully!\n\n";
    
    echo "📖 Key Takeaways:\n";
    echo "   • Use Client() to initialize the NodeMaven client\n";
    echo "   • Call getProxyConfig() with targeting options\n";
    echo "   • Use session IDs for sticky sessions\n";
    echo "   • ProxyConfig provides multiple integration methods\n";
    echo "   • Always test your proxy configurations\n\n";
    
    echo "🔗 Next Steps:\n";
    echo "   • Check proxy_integration.php for advanced usage\n";
    echo "   • Visit https://dashboard.nodemaven.com for account management\n";
    echo "   • Read the API documentation for more options\n";

} catch (NodeMavenException $e) {
    echo "❌ NodeMaven Error: " . $e->getMessage() . "\n";
    echo "💡 Status Code: " . $e->getStatusCode() . "\n";
    
    if ($e->getStatusCode() === 401 || $e->getStatusCode() === 403) {
        echo "💡 Check your API key in .env file or environment variables\n";
    }
    
} catch (Exception $e) {
    echo "❌ Unexpected Error: " . $e->getMessage() . "\n";
    echo "💡 Please check your configuration and try again\n";
} 